### PR TITLE
Add ContainerRuntimeMessage.compatDetails to futureproof against future "optional" ops

### PR DESCRIPTION
**Pull request by markfields**
Fixes [AB#4416](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4416), in preparation of adding GC "reference" and / or "deleted" ops - [AB#4168](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4168).

## Background

Currently, if a new container runtime op type is added and sent, it will crash older container runtimes because they will hit [this](https://github.com/microsoft/FluidFramework/blob/b294c000fafcf7b374e13ac0b81605dc44fef829/packages/runtime/container-runtime/src/containerRuntime.ts#L2002) assert. This makes adding new op types very time consuming. It has to be done in 2 stages - First, ship code that can read the new op type and once that saturates, ship code that writes the new op type.

## Description

This PR adds the ability for container runtime to not fail on receiving unknown op type but rather treating it as no-ops. This will enable adding new op types faster and it can be decided on a feature by feature basis to whether to wait for saturation or not. For example, the GC reference and deleted ops do not require saturation to be used.

@agarwal-navin is largely to thank for the design, which is to create a new optional `compatDetails` object on `ContainerRuntimeMessage` (the unpacked op). For now, it just has a `behavior` property which is either `Ignore` or `FailToProcess`. But being an object (not the string alone), it is extensible, in case variable data is needed to determine compatibility for some future `behavior` type.

## Reviewer Guidance

My biggest concern here is whether I covered all the codepaths where a `ContainerRuntimeMessage` is handled (e.g. serialized/deserialized). So I'm trying hard to get good code coverage - of course UT coverage of my changes, but more importantly coverage of existing op processing codepaths via e2e tests and stress tests so I can inject a "future" op and see that it works ok.

What I want to avoid is 6 months from now we introduce an "Ignored" op type and find out in production that I missed a spot.

### Follow-up To-Do's:

See [AB#5316](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/5316)

- [ ] Get coverage of round-tripping an op with compatDetails through batching/compression/chunking
- [ ] Get coverage of round-tripping an op with compatDetails through rebasing (op reentrancy)
- [ ] Get coverage of round-tripping an op with compatDetails through rollback (order sequentially) [low pri]
- [ ] Add back stress test changes and run them locally to validate it's working as expected

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:245]: Maybe too early to ask, but will this be part of the protocol (ie the server will relay it) or will we pack it further in the content?

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2065]: Is this differentiation valuable or should we just check for `Ignore`?

File: packages/runtime/container-runtime/src/pendingStateManager.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:247]: @kian-thompson / @wes-carlson - When I added in the new `compatDetails` property, JSON.stringify serialized the props in a different order here compared to `pendingMessage.content`.  I got kind of nervous about doing that comparison here, although I guess it's working?

Looks like it was added in response to [this conversation](https://github.com/microsoft/FluidFramework/pull/13994/files#r1100677930), how do we know the order of props will be preserved on server roundtrip?  Is it reliable enough that I can just be sure to serialize the same way here to get it to pass the tests, and it'll be find in prod?

I put a Draft PR out for a helper to do a proper deep comparison of the objects (#16834), but it's probably slower than this check - although, I also wonder how slow this check is already, this is on the critical path, right?

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:251]: Nope, it's buried inside op content

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2233]: Great observation.  I traced through the code and all callers catch and close the container.  I could have missed something but I feel pretty confident.

I'm open to adding back the explicit close, I think I removed it because I wanted to use `unreachableCase` which throws and it was awkward to try to close the container right here with that error.

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: If we add a new type to that enum, I want this switch statement to break until a case is added for it.  I refactored it, lmk how it looks now.

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2057]: Refactored it, take a look

File: packages/runtime/container-runtime/src/test/pendingStateManager.spec.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/test/pendingStateManager.spec.ts:417]: Previously `pendingStateManager` was any. Now this code knows `getLocalState()` could be undefined (but it won't be)

File: packages/test/test-end-to-end-tests/src/test/container.spec.ts
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/container.spec.ts:61]: (I will resolve these `//*` comments before merging)

File: packages/test/test-service-load/testConfig.json
Comment (markfields) [inline packages/test/test-service-load/testConfig.json:6]: Feedback on this number?  Maybe I'll make it smaller, like 5% instead of 10%

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (kian-thompson) [inline packages/runtime/container-runtime/src/containerRuntime.ts:238]: Not sure how I feel about this, but probably fine to revisit as other types are added

**Review by kian-thompson — APPROVED** [review]

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (andre4i) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2201]: I believe you are correct, but a test would be best.

**Review by andre4i — APPROVED** [review]
LGTM pending the test todos called out in the comments

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:238]: Here's the reason not to use the type here: The logic right now only calls out "Ignore"; the behavior for "FailToProcess" is implicit.  If another behavior is added, and this function is overlooked (unlikely, I know, since it's kind of the main point, but still) then the new option will default to the same behavior.  By explicitly calling out here what individual behaviors this function handles, if the other type changes it will force the author to deal with it, because the types will no longer be compatible.

It's similar to putting `unreachableCase` in the `default` case of a `switch` block that covers all known inputs.

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2201]: Btw - To generalize this across much of our code, I think it should be rare to have code originating an error also explicitly close the container.  If that's necessary, then what will `assert` do? Better to have our top-level APIs collect any error thrown from below and do the right thing.

**Pull request merged**
